### ISSUE TITLE
Make model learning and prediction explicit via the arguments

### ISSUE THREAD

**Issue by usptact** [bigartm/bigartm#433]
Hi,

I am using the CLI variant of the tool to learn the, the term-topic matrix (interpreting the topics), the document-topic matrix (classifying the documents) and some experiments in semi-supervised setting.

Perhaps I am missing a distinction between training and prediction, does it make sense to use explicitly commands like "train" and "predict" when invoking the tool?

Why I am asking this? I created a large dataset where some documents are labeled and some unlabeled. I ran the tool on this dataset to build the model. It took few hours but I got the model file. As the next step, I wish to get the label predictions for the unlabeled documents (I also keep them in a separate file). When I run the tool, I ask to load the unlabeled dataset as well as the trained model. It is still running but it appears that prediction (?) will run again those iterations as in training. Shouldn't it be faster? Am I missing something?

In summary, my point is to make training and prediction commands explicit in the tool.

**Comment by sashafrey** [comment]
Let's do it step by step - first, make sure you have your input data in a correct format, and then we can help to write down a specific commands for BigARTM CLI. Then we have a complete picture and can revisit BigARTM CLI to make it more intuitive. 

Would it be possible for you to share a snippet of your two input files in the VW-format (the one for training the model, and the other one without labels for classification)?

**Comment by usptact** [comment]
Sure, here are some typical examples of documents that I have.

The labeled ones:
d_10 |u call jim on his cell phone |class Calling
d_20 |u create an alarm for 5 p m |class Alarm
d_30 |u this sentence is false |class Other

The unlabeled ones:
u_10 |u dial john's mobile
u_20 |u give me an alarm for 8am in the morning
u_30 |u the sky is blue and the grass is green

During training, I mixed the two datasets together and shuffled it (unsure if it is needed). I asked for 100 topics (sparse phi and theta) + 10 noise (smooth phi and theta). I need to get back to the machine to provide you the exact parameters I specified.

The number of labeled examples: ~100K.
The number of unlabeled examples: ~32M.

**Comment by sashafrey** [comment]
1. I would try to build both models (with mixing the datasets, and just training on the labeled data). Overall mixing them sounds like a good idea - in my opinion it should help producing better topics for modality "u", and hence implicitly improve topics for modality "class".

2. The commands for CLI may look as follows:
```
  bigartm --read-vw-corpus labeled.txt --save-batches ~/labeled --save-dictionary labeled.dict
  bigartm --read-vw-corpus unlabeled.txt --save-batches ~/unlabeled
  bigartm --use-batches ~/labeled --use-dictionary labeled.dict --use-modality u,class
          --topics obj:100,background:10  --passes 10
          --regularizer "0.05 SparsePhi #obj"
          --regularizer "0.05 SparseTheta #obj"
          --regularizer "0.25 SmoothPhi #background"
          --regularizer "0.25 SmoothTheta #background"
          --save-model model.bin 
  bigartm --use-batches ~/unlabeled --use-modality u,class --topics 50
          --load-model model.bin
          --write-predictions unlabeled.pred.txt --csv-separator=tab
          --predict-class class --write-class-predictions unlabeled.pred_class.txt
          --score ClassPrecision
```
3. To classify I would suggest to split test data into subsets, and classify them let's say 1M at a time. You can do it just by moving batches into subfolders, and then specify the subfolder with fewer batches in `--use-batches` argument. One reason is that currently  CLI is stupid and for --write-predictions it stores and the entire Theta matrix in RAM. For your settings it would take 13 GB or RAM... I can fix this if needed. On another hand, it might be also hard to parse such a large prediction file. So by splitting into smaller changes it might be easier to consume the results on your side.

4 (last but not least) In the labeled examples rows correspond to the documents. To implement term-term matrix (e.g. the idea suggested by @AnyaP ) you would need to transform this data into something like this:
```
ITEM_call |u call jim on his cell phone |class Calling
ITEM_jim |u call jim on his cell phone |class Calling
ITEM_on |u call jim on his cell phone |class Calling
ITEM_his |u call jim on his cell phone |class Calling
ITEM_phone |u call jim on his cell phone |class Calling
ITEM_create |u create an alarm for 5 p m |class Alarm
ITEM_an |u create an alarm for 5 p m |class Alarm
ITEM_alarm |u create an alarm for 5 p m |class Alarm |u give me an alarm for 8am in the morning
...
```
BigARTM will treat rows as documents --- same as before. However, we cheat and create one row per WORD. There will be as many "documents" in this dataset as there are distinct tokens across your origina documents. Then, to create a row for a term you concatenate the documents where the term occurred. For example, word `alarm` was in `d_20` and `u_20`, so I put them together in `ITEM_alarm` (last line from my example). 

You can use **exactly the same commands** as I listed above to learn the model against this datasets and use this model to produce predictions. Obviously you need to change the name of the input vw-file (for example, `term.labeled.txt`), the folder to place batches (let's say `~/term.labeled`),
and save the model under a different name (`term.model.bin`). But most importantly, in the last command you should use `term.model.bin` exactly as `model.bin`.

@AnyaP What should we do with class labels in term-term representation of the collection?

**Comment by usptact** [comment]
Thanks! I will re-work my scripts to do the training as suggested. It is indeed faster to save the corpus in a binary format for subsequent processing. This allows to experiment more.

For the term-term approach, I believe I get the idea. But wouldn't that make an explosion of number of lines and the length of each line? There are many documents in the datasets I have (multiply by the average number of words) and many words are shared across the documents (multiply by the number of occurences). Wouldn't that make the dataset huge?

In my experiment I was mixing the labeled and unlabeled datasets in hope to run the semi-supervised learning.The hope is that those large albeit noisy datasets contain some useful information for this feature poor learning setting. Strictly speaking, this is transductive learning (some subtle differences with semi-supervised learning).

Yes, I noticed that prediction on the large unlabeled dataset was taking just too long. I killed it and will try to re run it in chunks.

Finally, I just obtained some sensible predictions in the semi-supervised setting for a smaller sample of data. Hopefully that for the large dataset (and more topics?) I will get decent classification performance.

Thanks for taking your time creating this detailed explanation with examples and other useful considerations. I will stay tuned!

**Comment by sashafrey** [comment]
> wouldn't that make an explosion of number of lines and the length of each line?
It should not explode because the number of lines will match the number of *distinct* tokens (not the total number of tokens). How many distinct tokens do you have across the collection? I call it *dictionary size*, and denote it `|W|`. For the collection of approx. your size this parameter should be around 100K or 300K.  To check this you may use BigARTM's command:
```
bigartm --use-batches ~/labeled --save-dictionary labeled.dict  --write-dictionary-readable labeled.dict.txt
```
The resulting file `labeled.dict.txt` will list all distinct tokens from the collection, and for each column tell you term frequency (tf) and document frequency (df). If there are too many tokens (>500K) you may filter them as in the following example:
```
  bigartm --use-dictionary labeled.dict --dictionary-min-df 5 dictionary-max-df 40% --save-dictionary labeled.filter.dict --write-dictinoary-readable labeled.filter.dict.txt
```
You may parse the `labeled.filter.dict.txt` and use only these tokens in your script where you generate training dataset (ignore tokens with too low or too high term frequency).

**Comment by usptact** [comment]
Makes sense. The currently running bigartm process provided the following line:
Number of tokens in the model: 215162

Is this the number of unique terms found in my corpus? I mixed my labeled ~100K, the unlabeled ~32M and an additional unlabeled corpus ~185K (contains many documents of one class of interest) into one big dataset. Few iterations more and I should get my model.

**Comment by sashafrey** [comment]
Yes, `215162` is the number of distinct tokens after filtering the dictionary with `min-df 3` and `max-df 40%`.

**Comment by AnyaP** [comment]
> But wouldn't that make an explosion of number of lines and the length of each line? 
The number of lines will not explode and will be equal to vocabulary size, as @sashafrey mentioned. Also, is it English or Russian? And do you use lemmatization? It would further reduce the vocabulary size and improve performance. I believe that in most cases up to 100 000 unique tokens should be enough (just look through those terms that have the lowest document frequency and decide on whether they make sense).

But you are absolutely right that the length of each line is likely to explode. In my experience it would result in a really huge file. So it would be really helpful if you could aggregate the counters (before feeding data to BigARTM). I mean write down in the VW-file something like wampinuk:3 instead of wampinuk wampinuk wampinuk. Then the file shouldn't be larger then the initial one and it will be processed pretty fast as opposite to the unaggregated data. Also, I would recommend to filter out those word co-occurrence that are low (i.e. equal to 1). If two words has co-occurred only once among 100K(or more) documents, then it is likely to be occasional.

By the way, it's easy to estimate that word-word representation in non-aggregated format will be 5-8 (average doc length) times larger than the initial corpus, because each document will be copied for each word in it. So you can decide whether it's acceptable. Aggregated format is much more scalable because starting from some point, new documents do not nearly produce new non-zero word co-occurrences.

> @AnyaP What should we do with class labels in term-term representation of the collection?
Well, ideally we should fix Phi_1 matrix for text modality based on learning in term-term representation and then lean Phi_2 matrix for class labels and Theta for documents based on standard representation. Is it possible to fix (or even better update more slowly in online mode) Phi for one modality and keep it changing for another?

**Comment by usptact** [comment]
Thanks, @AnyaP, for explanation on preparation of the dataset. The only downside to this representation is that it might be harder to add/remove documents. One might need to re-create it every time after the corpus has been modified. In my case the corpus can change although not very often so this is not a big problem.

Could you please explain little bit more the last paragraph? Why would you want to learn \Phi_1 on tterm-term data and then \Phi_2 for class labels + \Theta for documents? Why is it desireable to keep one modality fixed while another can change? Just curious about workings.

**Comment by AnyaP** [comment]
Well, we assume that:
1) Phi_1 (probabilities of words in topics) can suffer from sort texts, so it would be reasonable to use word-word instead of word-document data for learning them.
2) Word-word data doesn't contain class labels for documents, just because there are no documents in that case :) So we can't learn Phi_2 (probabilities of classes for topics) on that data. One could think about mixing words and labels in word-word data to have a bigger matrix, however I've never tried it and can't say if it works.

Thus, I suggest that we should learn Phi_1 on word-word data and Phi_2 and Theta on word-document data. We could also try to adjust Phi_1 on word-document data, but not rewrite it completely, otherwise having two steps makes no sense.

**Comment by sashafrey** [comment]
@usptact I'm moving this ticket to the backlog because in this ticket I don't see any specific actions that needs to be done on BigARTM side. Except, of course, a better documentation -- hence it goes to the backlog, and we will revisit it as soon as possible.

Please let us know if we can help further with this question.
